odoo/addons/base/security/base_security.xml

```diff
@@ -152,6 +155,14 @@
             <field name="domain_force">['|', ('share', '=', False), ('company_ids', 'in', company_ids)]</field>
         </record>
 
+        <record id="res_users_company_consistency" model="ir.rule">
+            <field name="name">user company_consistency</field>
+            <field name="model_id" ref="model_res_users"/>
+            <field name="domain_force">[('company_ids', 'in', company_ids)]</field>
+            <field name="company_consistency" eval="1"/>
+            <field name="active" eval="0"/>
```

Comment:
No, because otherwise you wouldn't be able to see all internal users like the previous rule suggests.
This one basically only manages the `_check_company` (and the domain in the UI) between related records; but it won't prevent seeing the internal user generally. (you can still talk to them etc.)